### PR TITLE
mon/MDSMonitor: fix standby-replay mds being removed from MDSMap unexpectedly

### PR DESCRIPTION
**Pull request by huww98**
mds/FSMap: stricter state_transition_valid

Reject any unknown transitions.

mon/MDSMonitor: remove redundant state change check

There are two sets of checks to state change in prepare_beacon. Since the last commit, the first set is all covered by
`MDSMap::state_transition_valid`. So remove the first set and move the second set to its place.

And some misc fixes.

Fixes: https://tracker.ceph.com/issues/53811

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/cephfs/.gitignore
Comment (batrick) [inline doc/cephfs/.gitignore:1]: :+1:

File: doc/cephfs/mds-state-diagram.dot
Comment (batrick) [inline doc/cephfs/mds-state-diagram.dot:60]: :+1:

File: src/mds/MDSMap.cc
Comment (batrick) [inline src/mds/MDSMap.cc:1019]: Looks like we are also missing a state transition line for REJOIN -> STOPPED. Mind updating that one too?

**Review by batrick — CHANGES_REQUESTED** [review]
Great changes, thank you! Just a small doc update requested.

File: src/mds/MDSMap.cc
Comment (huww98) [reply src/mds/MDSMap.cc:1019]: OK

**Comment by huww98** [comment]
I initialize `MDSRank::state` to standby in this revision

**Comment by huww98** [comment]
jenkins test make check arm64

File: src/mds/MDSRank.h
Comment (batrick) [inline src/mds/MDSRank.h:412]: \
```suggestion
    MDSMap::DaemonState state = MDSMap::STATE_NULL;
```

would be more appropriate I think.

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise LGTM

File: src/mds/MDSRank.h
Comment (huww98) [reply src/mds/MDSRank.h:412]: if it is not `MDSMap::STATE_STANDBY`, the [added assert](https://github.com/ceph/ceph/pull/44501/files#diff-f91675b3e095c571fa2e401547ac27a8f0b1575788f0df3321a8f20a5949f90dR2242) would fail, or the [`MDSMap::state_transition_valid(oldstate, state)`](https://github.com/ceph/ceph/pull/44501/files#diff-f91675b3e095c571fa2e401547ac27a8f0b1575788f0df3321a8f20a5949f90dR2261) would return false.

Comment (vshankar) [reply src/mds/MDSRank.h:412]: I think @batrick is right - This causes an failed assert::

```
    -5> 2022-02-05T21:48:44.562+0000 7fc710419700  4 mds.0.0 apply_blocklist: killed 0 blocklisted sessions (0 blocklist entries, 0)
    -4> 2022-02-05T21:48:44.562+0000 7fc710419700 10 mds.c handle_mds_map: handling map as rank 0
    -3> 2022-02-05T21:48:44.562+0000 7fc710419700 -1 /build/ceph-17.0.0-10531-g09fc2e9e/src/mds/MDSRank.cc: In function 'void MDSRankDispatcher::handle_mds_map(ceph::cref_t<MMDSMap>&, const MDSMap&)' thread 7fc710419700 time 2022-02-05T21:48:44.566988+0000
/build/ceph-17.0.0-10531-g09fc2e9e/src/mds/MDSRank.cc: 2242: FAILED ceph_assert(state == oldmap.get_state_gid(mds_gid))

 ceph version 17.0.0-10531-g09fc2e9e (09fc2e9e470435f1e6dbf462a51853391f4a4cc0) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14f) [0x7fc71593c7cc]
 2: /usr/lib/ceph/libceph-common.so.2(+0x27b9de) [0x7fc71593c9de]
 3: (MDSRankDispatcher::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&, MDSMap const&)+0x2df7) [0x7fc716684dc7]
 4: (MDSDaemon::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&)+0x107b) [0x7fc71665b67b]
 5: (MDSDaemon::handle_core_message(boost::intrusive_ptr<Message const> const&)+0x3df) [0x7fc71665d1ef]
 6: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x186) [0x7fc71665db86]
 7: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x460) [0x7fc715bd48d0]
 8: (DispatchQueue::entry()+0x58f) [0x7fc715bd216f]
 9: (DispatchQueue::DispatchThread::entry()+0x11) [0x7fc715ca0441]
 10: /lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7fc71566c609]
 11: clone()
```

Happens when an MDSRank gets initialized for the very first time - https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6663006/

(You may want to vet this with your other changes in the PR).

**Review by vshankar — CHANGES_REQUESTED** [review]
PTAL at the failed test.

File: src/mds/MDSRank.h
Comment (huww98) [reply src/mds/MDSRank.h:412]: This failing assert is added in this PR.

I assumed MDSRank would only be initialized when the MDSDaemon is in STANDBY state. This should be true according to the [state diagram](https://docs.ceph.com/en/pacific/cephfs/mds-states/#lines). After the changes in this PR, MDS should reject any state change that is not in the diagram.

I will take a look. Do you have any idea why the MDS would initialize MDSRank while not in STANDBY state? Or how can I reproduce this locally?

Or maybe I should just relax the assert, and allow MDSRank to be initialized in any state.

Comment (huww98) [reply src/mds/MDSRank.h:412]: OK, I've managed to reproduce this by setting `paxos_propose_interval` to 5.0 then restarting MDS. Will fix this tomorrow.

Comment (vshankar) [reply src/mds/MDSRank.h:412]: thx @huww98

**Comment by vshankar** [comment]
@huww98 ping?

**Comment by huww98** [comment]
Just added a branch so that MDS would not panic if it does not get the STANDBY map.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
jenkins test api

**Comment by huww98** [comment]
jenkins test api

**Comment by huww98** [comment]
ping. @vshankar @batrick . Anything blocking this PR? jenkins tests all passed now.

**Comment by vshankar** [comment]
> ping. @vshankar @batrick . Anything blocking this PR? jenkins tests all passed now.

I was off last week - will pick this up in my next batch.

**Review by batrick — APPROVED** [review]
lgtm

**Comment by batrick** [comment]
@huww98 apologies! I forgot about this fix you wrote when I myself stumbled on the problem. Your other changes are still good, can you rebase and we'll include in the next test branch.

**Comment by huww98** [comment]
Hi @batrick . I just resolve the conflict by overwriting your changes in b2c40cc07faa, since these checks should already be covered by the updated `MDSMap::state_transition_valid`

Also, I refined the following check in this version to better match the original logic.

```diff
@@ -697,9 +706,11 @@ bool MDSMonitor::prepare_beacon(MonOpRequestRef op)
     }
 
     // legal state change?
-    if (info.state == MDSMap::STATE_STANDBY && state != info.state) {
-      // Standby daemons should never modify their own
-      // state.  Reject any attempts to do so.
+    if ((info.state == MDSMap::STATE_STANDBY && state != info.state) ||
+       (info.state == MDSMap::STATE_STANDBY_REPLAY && state != info.state && state != MDSMap::STATE_DAMAGED)) {
+      // Standby daemons should never modify their own state.
+      // Except that standby-replay can indicate the rank is damaged due to failure to replay.
+      // Reject any attempts to do so.
       derr << "standby " << gid << " attempted to change state to "
            << ceph_mds_state_name(state) << ", rejecting" << dendl;
       goto evict;
```

**Comment by batrick** [comment]
Yes, that adjustment looks good!

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test api

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@huww98 please do a rebase.

**Comment by huww98** [comment]
rebased

**Comment by batrick** [comment]
retest this please

**Comment by cbodley** [comment]
jenkins test make check

**Comment by vshankar** [comment]
I'll pick this up in my next test run (sometime this week).

**Comment by batrick** [comment]
```
 23/284 Test #271: run-tox-jsonnet-lint ......................***Failed    7.66 sec
Requirement already satisfied: tox in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (3.25.1)
Requirement already satisfied: six>=1.14.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (1.16.0)
Requirement already satisfied: packaging>=14 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (21.3)
Requirement already satisfied: pluggy>=0.12.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (1.0.0)
Requirement already satisfied: virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (20.16.3)
Requirement already satisfied: filelock>=3.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (3.8.0)
Requirement already satisfied: toml>=0.9.4 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (0.10.2)
Requirement already satisfied: py>=1.4.17 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from tox) (1.10.0)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from packaging>=14->tox) (2.4.7)
Requirement already satisfied: distlib<1,>=0.3.5 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (0.3.5)
Requirement already satisfied: platformdirs<3,>=2.4 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.10/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (2.5.2)
jsonnet-lint create: /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/.tox/jsonnet-lint
jsonnet-lint installdeps: -rrequirements-grafonnet.txt
WARNING: Discarding $PYTHONPATH from environment, to override specify PYTHONPATH in 'passenv' in your configuration.
jsonnet-lint installed: jsondiff==2.0.0
jsonnet-lint run-test-pre: PYTHONHASHSEED='4162932440'
jsonnet-lint run-test: commands[0] | ./lint-jsonnet.sh --test
/home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh: 5: jsonnetfmt: not found
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh --test (exited with code 127)
```

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
jenkins test api

**Comment by vshankar** [comment]
@huww98 @batrick I'm seeing a test failure: https://pulpito.ceph.com/vshankar-2022-08-18_04:30:42-fs-wip-vshankar-testing1-20220818-082047-testing-default-smithi/6978338/

```
"2022-08-18T10:17:37.330342+0000 mon.a (mon.0) 522 : cluster [WRN] Replacing daemon mds.b as rank 0 with standby daemon mds.e" in cluster log
```

Doesn't look like its caused by this change, but I'll like to confirm before merging.

**Comment by vshankar** [comment]
FWIW, I do not see anything unusual in the MDS log, but the monitors decided to kick out the MDSs for some reason.

**Comment by batrick** [comment]
https://github.com/ceph/ceph/pull/47754

**Comment by vshankar** [comment]
jenkins test make check arm64

**Review by vshankar — APPROVED** [review]

**Pull request merged**
